Propagate timeout errors up in fill phase.

baldersheim opened the pull request:
@bratseth PR

baldersheim committed:
Propagate timeout errors up in fill phase.

bratseth commented on container-search/src/main/java/com/yahoo/search/dispatch/Dispatcher.java:
```diff
@@ -222,12 +224,28 @@ private void processResponse(Client.GetDocsumsResponseOrError responseOrError) {
+            errors.traverse((ArrayTraverser)(int index, com.yahoo.slime.Inspector value) -> {
+                ByteArrayOutputStream os = new ByteArrayOutputStream(1024);
+                try {
+                    new JsonFormat(true).encode(os, value);
```
How about we define the content of this instead of passing JSON through?
I suggest a mandatory "type" and "message" field and an optional "details" field, as that can be mapped cleanly to ErrorMessage.

baldersheim replied:
Done.

bratseth commented on container-search/src/main/java/com/yahoo/search/dispatch/Dispatcher.java:
```diff
@@ -17,12 +17,14 @@
-import com.yahoo.slime.BinaryFormat;
-import com.yahoo.slime.Cursor;
-import com.yahoo.slime.Slime;
+import com.yahoo.slime.*;
```
No * imports please.

baldersheim replied:
New mac new settings...

bratseth commented on container-search/src/main/java/com/yahoo/search/dispatch/Dispatcher.java:
```diff
@@ -222,12 +224,28 @@ private void processResponse(Client.GetDocsumsResponseOrError responseOrError) {
-            Inspector summaries = new SlimeAdapter(BinaryFormat.decode(slimeBytes).get().field("docsums"));
+            com.yahoo.slime.Inspector root = BinaryFormat.decode(slimeBytes).get();
+            Inspector summaries = new SlimeAdapter(root.field("docsums"));
             if ( ! summaries.valid())
```
This will lose the error information if we ever return only errors.
It would be (slightly) more robust to parse errors first and then summaries, and only fail the latter if there are no errors already.

bratseth commented on container-search/src/main/java/com/yahoo/prelude/fastsearch/VespaBackEndSearcher.java:
```diff
@@ -500,11 +504,21 @@ protected int fillHits(Result result, int packetIndex, Packet[] packets, String
+                try {
+                    if (!fillHit(fastHit, docsum, summaryClass))
+                        skippedHits++;
+                } catch (TimeoutException e) {
```
Using an exception to propagate this may not be very good in the scenario where someone asks for very many hits, thus causing a timeout, or if the container is already overloaded, causing some timeouts.
I suggest changing the return value of fillHit to be e.g a FillResult object containing 2 fields
boolean success; // otherwise skipped aka not available
Optional<String> error;

baldersheim replied:
Agree.

havardpe commented on container-search/src/main/java/com/yahoo/prelude/fastsearch/DocsumDefinitionSet.java:
```diff
@@ -57,7 +60,7 @@ public final DocsumDefinition getDocsumDefinition(String name) {
      * @throws ConfigurationException if the summary class of this hit is missing
      */
-    public final void lazyDecode(String summaryClass, byte[] data, FastHit hit) {
+    public final String lazyDecode(String summaryClass, byte[] data, FastHit hit) {
```
please add a description of the return value in the comment

havardpe reviewed (commented):
I agree with Jon that exceptions might not be the optimal way to propagate this. Also, error messages should be somewhat structured. This is what I have so far for slime-based message passing (could be releavnt in this context. Note that error types are strings, not numbers):

request: ["method", [parameter*]]
reply: [[error*], [return value*]]
error: { "type": "string", "message": "string", "context": whatever }

baldersheim committed:
Use a ResultObject to avoid using Exception for value passing.

baldersheim committed:
Avoid import by *.
Restructure error parsing to avoid loosing information when you only have errors.
Use a fixed mandatory "type", "message" "details" schema for reporting errors.

baldersheim committed:
Add comment on return value

baldersheim committed:
Add comment on return value

baldersheim commented:
"type", "message", "details". All strings.

baldersheim commented:
@bratseth Ready for another review

bratseth commented on container-search/src/main/java/com/yahoo/search/dispatch/Dispatcher.java:
```diff
@@ -221,13 +226,37 @@ private void processResponse(Client.GetDocsumsResponseOrError responseOrError) {
 
+        private void addErrors(com.yahoo.slime.Inspector errors) {
+            errors.traverse((ArrayTraverser) (int index, com.yahoo.slime.Inspector value) -> {
+                ByteArrayOutputStream os = new ByteArrayOutputStream(1024);
```
I think you can remove 231-236 now?

baldersheim replied:
Yes, can do, done

bratseth reviewed (commented):
One question, otherwise LGTM!

baldersheim committed:
remove unused code
